File: solucion JS/public/js/game.js
// Diccionario Español-Inglés:
// fila: row, columna: col, tablero: board, pista: hint, solucion: solution, inicial: initial, vacía: empty, numero: number, completo: completed

class SudokuGame {
  constructor(initial, solution) {
    this.initial = initial.map(row => row.slice());
    this.solution = solution.map(row => row.slice());
    this.board = initial.map(row => row.slice());
  }

  isValidMove(row, col, num) {
    if (this.initial[row][col] !== 0) return false;
    if (!(num >= 1 && num <= 9)) return false;
    for (let i = 0; i < 9; i++) {
      if (this.board[row][i] === num && i !== col) return false;
      if (this.board[i][col] === num && i !== row) return false;
    }
    let startRow = 3 * Math.floor(row / 3), startCol = 3 * Math.floor(col / 3);
    for (let i = startRow; i < startRow + 3; i++)
      for (let j = startCol; j < startCol + 3; j++)
        if (this.board[i][j] === num && (i !== row || j !== col)) return false;
    return true;
  }

  isCompleted() {
    for (let i = 0; i < 9; i++)
      for (let j = 0; j < 9; j++)
        if (this.board[i][j] !== this.solution[i][j]) return false;
    return true;
  }

  getRandomHint() {
    let empty = [];
    for (let i = 0; i < 9; i++)
      for (let j = 0; j < 9; j++)
        if (this.board[i][j] === 0) empty.push([i, j]);
    if (!empty.length) return null;
    let [i, j] = empty[Math.floor(Math.random() * empty.length)];
    return { row: i, col: j, num: this.solution[i][j] };
  }
}

File: solucion Python/logic/game.py
import random

class SudokuGame:
    def __init__(self, initial, solution):
        # Tablero inicial y solución (listas de listas)
        self.initial = [row[:] for row in initial]
        self.solution = [row[:] for row in solution]
        self.board = [row[:] for row in initial]  # Estado actual del tablero

    def is_valid_move(self, row, col, num):
        # Verifica si se puede poner 'num' en (row, col)
        if self.initial[row][col] != 0:
            return False
        if not (1 <= num <= 9):
            return False
        for i in range(9):
            if self.board[row][i] == num and i != col:
                return False
            if self.board[i][col] == num and i != row:
                return False
        start_row, start_col = 3*(row//3), 3*(col//3)
        for i in range(start_row, start_row+3):
            for j in range(start_col, start_col+3):
                if self.board[i][j] == num and (i, j) != (row, col):
                    return False
        return True

    def is_completed(self):
        # Verifica si el tablero está completo y correcto
        for i in range(9):
            for j in range(9):
                if self.board[i][j] != self.solution[i][j]:
                    return False
        return True

    def get_random_hint(self):
        # Devuelve una celda vacía aleatoria y su valor correcto
        empty = [(i, j) for i in range(9) for j in range(9) if self.board[i][j] == 0]
        if not empty:
            return None
        i, j = random.choice(empty)
        return (i, j, self.solution[i][j])

"""
- fila: row
- columna: col / column
- tablero: board
- pista: hint
- solucion: solution
- inicial: initial
- vacía: empty
- numero: number
- completo: completed
"""